cumulus/client/cirrus-executor/src/unsigned_submitter.rs

```diff
@@ -0,0 +1,111 @@
+use crate::LOG_TARGET;
+use futures::FutureExt;
+use sc_client_api::HeaderBackend;
+use sp_api::ProvideRuntimeApi;
+use sp_core::traits::SpawnNamed;
+use sp_executor::{BundleEquivocationProof, ExecutorApi, FraudProof, InvalidTransactionProof};
+use sp_runtime::generic::BlockId;
+use sp_runtime::traits::Block as BlockT;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::error::TrySendError;
+use tokio::sync::mpsc::Sender;
+
+const UNSIGNED_MESSAGE_BUFFER_SIZE: usize = 128;
+
+#[derive(Debug, Clone)]
+pub(crate) enum UnsignedMessage {
+    Fraud(FraudProof),
+    BundleEquivocation(BundleEquivocationProof),
+    InvalidTransaction(InvalidTransactionProof),
+}
+
+/// Submits various executor-specific unsigned extrinsic to the primary node.
+#[derive(Clone)]
+pub(crate) struct UnsignedSubmitter {
+    sender: Sender<UnsignedMessage>,
+}
+
+impl UnsignedSubmitter {
+    pub(crate) fn new<Block, PBlock, PClient>(
+        primary_chain_client: Arc<PClient>,
+        spawner: Box<dyn SpawnNamed + Send + Sync>,
+    ) -> Self
+    where
+        Block: BlockT,
+        PBlock: BlockT,
+        PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + 'static,
+        PClient::Api: ExecutorApi<PBlock, Block::Hash>,
```

Comment:
Doesn't this work too?:
```suggestion

    pub(crate) fn new<Block, PBlock, PClient, Spawner>(
        primary_chain_client: Arc<PClient>,
        spawner: &Spawner,
    ) -> Self
    where
        Block: BlockT,
        PBlock: BlockT,
        PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + 'static,
        PClient::Api: ExecutorApi<PBlock, Block::Hash>, 
        Spawner: SpawnNamed,
```

It doesn't seem to be necessary to move it anywhere or own.

Reply:
I didn't try to compile it, but this simple change compiles:
```patch
Index: cumulus/client/cirrus-executor/src/lib.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/cumulus/client/cirrus-executor/src/lib.rs b/cumulus/client/cirrus-executor/src/lib.rs
--- a/cumulus/client/cirrus-executor/src/lib.rs	(revision 9d6e45856c2f86fe9bd522b7ebbe9bbdc9dd0e1c)
+++ b/cumulus/client/cirrus-executor/src/lib.rs	(date 1658337149847)
@@ -229,9 +229,9 @@
             code_executor,
         );
 
-        let unsigned_submitter = UnsignedSubmitter::new::<Block, PBlock, PClient>(
+        let unsigned_submitter = UnsignedSubmitter::new::<Block, PBlock, PClient, _>(
             primary_chain_client.clone(),
-            spawner.clone(),
+            spawner.as_ref(),
         );
 
         let bundle_processor = BundleProcessor::new(
Index: cumulus/client/cirrus-executor/src/unsigned_submitter.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/cumulus/client/cirrus-executor/src/unsigned_submitter.rs b/cumulus/client/cirrus-executor/src/unsigned_submitter.rs
--- a/cumulus/client/cirrus-executor/src/unsigned_submitter.rs	(revision 9d6e45856c2f86fe9bd522b7ebbe9bbdc9dd0e1c)
+++ b/cumulus/client/cirrus-executor/src/unsigned_submitter.rs	(date 1658337109536)
@@ -27,15 +27,16 @@
 }
 
 impl UnsignedSubmitter {
-    pub(crate) fn new<Block, PBlock, PClient>(
+    pub(crate) fn new<Block, PBlock, PClient, Spawner>(
         primary_chain_client: Arc<PClient>,
-        spawner: Box<dyn SpawnNamed + Send + Sync>,
+        spawner: &Spawner,
     ) -> Self
     where
         Block: BlockT,
         PBlock: BlockT,
         PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + 'static,
         PClient::Api: ExecutorApi<PBlock, Block::Hash>,
+        Spawner: SpawnNamed + ?Sized,
     {
         let (sender, mut receiver) = mpsc::channel(UNSIGNED_MESSAGE_BUFFER_SIZE);
```

Reply:
Hmm, looks like the trick is using `spawner.as_ref()`, I was using `&spawer` when calling UnsignedSubmitted::new(), the compiler is smart enough to lead me to the right place, thank you!